### PR TITLE
start cc docs guide

### PR DESCRIPTION
**Pull request by pngwn**
## Description

Please include a concise summary, in clear English, of the changes in this pull request. If it closes an issue, please mention it here.

Closes: #(issue)

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

File: guides/05_custom-components/09_documenting-custom-components.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+#### What do I need to add hints to?
+
+You do not need to add type hints to every part of your code. In order for the documentation to work correctly, you will need to add type hints to the following component methods:
+
+- `__init__` parameters should be typed.
+- `postprocess` parameters and return value should be typed.
+- `preprocess` parameters and return value should be typed.
+
+If you are using `gradio cc create` then these types should already be provided but you may need to tweak them based on anay changes you make.
```
Comment (freddyaboulton) [inline guides/05_custom-components/09_documenting-custom-components.md:63]: Typo: `anay`

File: guides/05_custom-components/09_documenting-custom-components.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+##### `__init__`
+
+Here you only need to type the parameters. If you have cloned a temaplte with `gradio cc create` then these should already be in place. You will only need to add new hints for anything you have added or change:
```
Comment (freddyaboulton) [inline guides/05_custom-components/09_documenting-custom-components.md:67]: Typo: `tempalte`

File: guides/05_custom-components/09_documenting-custom-components.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+```py
+def __init__(
```
Comment (freddyaboulton) [inline guides/05_custom-components/09_documenting-custom-components.md:70]: Maybe have an example with fewer union types to not overwhelm newcomers?

File: guides/05_custom-components/09_documenting-custom-components.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+### Docstrings
+
+Docstrings also used extensively to extract more meaningful, human-readable descriptions of certain parts of the API.
+
+<details> 
+<summary> What are docstrings?</summary>
+
+
+If you aren't familiar with docstrings in Python, they are a way to annotate parts of your code with human-readble desciptions and explanations. They offer a rich in-editor experience like type hints but unlike type hints, they don't have any specific syntax requirements. They are simple strings and can take almost any form. The only requirement is where they appear. Docstrings should be "a string literal that occurs as the first statement in a module, function, class, or method definition".
```
Comment (freddyaboulton) [inline guides/05_custom-components/09_documenting-custom-components.md:116]: Typo: `human-readble`

File: js/_website/src/lib/components/Details.svelte [hunk 5]
```diff
@@ -0,0 +1,52 @@
+<script lang="ts">
+	import Lightbulb from "$lib/components/icons/Lightbulb.svelte";
```
Comment (freddyaboulton) [inline js/_website/src/lib/components/Details.svelte:2]: Nice!

**Review by freddyaboulton — APPROVED** [review]
Looks great @pngwn ! Just noticed some typos. Would be good to add the docs step to the five minute guide.

**Comment by aliabd** [comment]
you might still be working on this but just a note that the details are not rendering properly. they go to the top and stay on the page even after navigating to a different guide.

Love them though

**Comment by abidlabs** [comment]
Why don't we make `gradio cc docs` a part of `gradio cc publish` that is automatically carried out by default? It makes the resulting published Spaces look a lot nicer so we should encourage it by making it part of the default workflow.

**Comment by pngwn** [comment]
@abidlabs We actually can't do that as @freddyaboulton and I found out on friday. Since `gradio cc docs` generates a markdown file that we want to display on PyPi (as well as other places), we need the docs generation to happen _before_ the build, because the README.md that is currently in the repo gets packaged into the wheel to be displayed on PyPi. . We can, and should, include this command as part of the build by default though.

The only reason we haven't done that is just to get the PR merged. I'll make a PR for integrating the command into `build`.

**Comment by abidlabs** [comment]
Thanks for clarifying @pngwn! My main concern is having too many commands for developers to keep track off, so including `gradio cc docs` within `gradio cc build` sounds good to me

**Comment by pngwn** [comment]
We will still have the docscommand in case people want more control but the build can have some nice defaults for docs.

File: guides/05_custom-components/09_documenting-custom-components.md (hunk 3 above)
Comment (pngwn) [reply guides/05_custom-components/09_documenting-custom-components.md:91]: Simplified this!

**Comment by pngwn** [comment]
Should be good to review now! Thanks for the comments.

File: .vscode/settings.json [hunk 6]
```diff
@@ -19,5 +19,12 @@
...
 	},
-	"typescript.tsdk": "node_modules/typescript/lib"
+	"typescript.tsdk": "node_modules/typescript/lib",
+	"grammarly.selectors": [
+		{
+			"language": "markdown",
+			"scheme": "file",
+			"pattern": "guides/05_custom-components/09_documenting-custom-components.md"
+		}
+	]
```
Comment (pngwn) [inline .vscode/settings.json:29]:
```suggestion
	"typescript.tsdk": "node_modules/typescript/lib"
```

File: guides/05_custom-components/09_documenting-custom-components.md [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+
+  EVENTS = [
+    Events.change,
+    Events.upload,
+  ]
+```
+
+#### Custom events
+
+You can define a custom event if the built-in events are unsuitable for your use case. This is a straightforward process, but you must create the event in this way for docstrings to work correctly:
```
Comment (freddyaboulton) [inline guides/05_custom-components/09_documenting-custom-components.md:204]: nice explanation 👍

**Review by freddyaboulton — APPROVED** [review]
Looks great @pngwn ! I think we should mention that docs get built in the build command in the five minute guide because that's how most people start building custom components and they may not make it to the later guides

**Comment by pngwn** [comment]
Ah 5 minute guide, I forgot that. Will add.

**Comment by pngwn** [comment]
Added @freddyaboulton!

**Pull request merged**